pkg/apis/serving/v1alpha1/revision_validation.go

```diff
@@ -185,119 +200,181 @@ func ValidateContainerConcurrency(cc RevisionContainerConcurrencyType, cm Revisi
 }
 
 func validateVolume(volume corev1.Volume) *apis.FieldError {
-	var errs *apis.FieldError
+	errs := validateDisallowedFields(volume, *serving.VolumeMask(&volume))
 	if volume.Name == "" {
 		errs = apis.ErrMissingField("name")
 	} else if len(validation.IsDNS1123Label(volume.Name)) != 0 {
 		errs = apis.ErrInvalidValue(volume.Name, "name")
 	}
 
 	vs := volume.VolumeSource
-	switch {
-	case vs.Secret != nil, vs.ConfigMap != nil:
-		// These are fine.
-	default:
+	errs = errs.Also(validateDisallowedFields(vs, *serving.VolumeSourceMask(&vs)))
+	if vs.Secret == nil && vs.ConfigMap == nil {
 		errs = errs.Also(apis.ErrMissingOneOf("secret", "configMap"))
 	}
 	return errs
 }
 
+func validateEnvValueFrom(source *corev1.EnvVarSource) *apis.FieldError {
+	if source == nil {
+		return nil
+	}
+	return validateDisallowedFields(*source, *serving.EnvVarSourceMask(source))
+}
+
+func validateEnv(envVars []corev1.EnvVar) *apis.FieldError {
+	var errs *apis.FieldError
+	for i, env := range envVars {
+		errs = errs.Also(validateDisallowedFields(env, *serving.EnvVarMask(&env)).ViaIndex(i)).Also(
+			validateEnvValueFrom(env.ValueFrom).ViaField("valueFrom").ViaIndex(i))
+	}
+	return errs
+}
+
+func validateEnvFrom(envFromList []corev1.EnvFromSource) *apis.FieldError {
```

Comment:
I think we want the validation up a level as well.

Reply:
Not sure I see what is missing here. I think I hit validation at all the levels here.

Reply:
I didn't understand your comment in chat earlier.  What I meant is that I think we should have shared validation logic across versions (like the field mask), so ideally these functions would just move to `pkg/apis/serving`.